### PR TITLE
Separate WAN config for built-in and custom publishers

### PR DESCRIPTION
**Pull request by mmedenjak**
The previously used wan-publisher (WanPublisherConfig) is now separated
into two config elements/classes: batch-publisher
(WanBatchReplicationPublisherConfig) and custom-publisher
(CustomWanPublisherConfig). This allows the built-in batch publisher to
be configured more easily with more IDE support guiding the user through
the configuration and documentation. At the same time, the custom
publisher configuration is simplified since all mentioning of queues
and internal implementation details have been removed.

Stringly typed properties were migrated to config class fields to allow
for documentation, better type constraints and understandability and to
avoid wrongly typing in the name of the property which silently renders
the config unused.

Previously:
```xml
<wan-publisher group-name="nyc" publisher-id="nycPublisherId">
    <class-name>com.hazelcast.enterprise.wan.impl.replication.WanBatchReplication</class-name>
    <queue-capacity>15000</queue-capacity>
    <queue-full-behavior>DISCARD_AFTER_MUTATION</queue-full-behavior>
    <initial-publisher-state>REPLICATING</initial-publisher-state>
    <wan-sync>
        <consistency-check-strategy>NONE</consistency-check-strategy>
    </wan-sync>
    <properties>
        <property name="endpoints">10.3.5.1:5701,10.3.5.2:5701</property>
        <property name="batch.size">1000</property>
        <property name="batch.max.delay.millis">2000</property>
        <property name="response.timeout.millis">60000</property>
        <property name="ack.type">ACK_ON_OPERATION_COMPLETE</property>
        <property name="snapshot.enabled">false</property>
        <property name="group.password">nyc-pass</property>
    </properties>
</wan-publisher>
```

Now:
```xml
<batch-publisher>
            <group-name>nyc</group-name>
            <publisher-id>nycPublisherId</publisher-id>
            <batch-size>1000</batch-size>
            <batch-max-delay-millis>2000</batch-max-delay-millis>
            <response-timeout-millis>60000</response-timeout-millis>
            <acknowledge-type>ACK_ON_OPERATION_COMPLETE</acknowledge-type>
            <initial-publisher-state>REPLICATING</initial-publisher-state>
            <snapshot-enabled>false</snapshot-enabled>
            <queue-full-behavior>DISCARD_AFTER_MUTATION</queue-full-behavior>
            <queue-capacity>10000</queue-capacity>
    <target-endpoints>10.3.5.1:5701,10.3.5.2:5701</target-endpoints>
    <wan-sync>
        <consistency-check-strategy>NONE</consistency-check-strategy>
    </wan-sync>
</batch-publisher>
```

EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3078

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/config/AbstractWanPublisherConfig.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+     * @param implementation the implementation for the WAN replication
+     * @return this config
+     */
+    public AbstractWanPublisherConfig setImplementation(Object implementation) {
+        this.implementation = implementation;
+        return this;
+    }
+
+    @Override
+    public int getFactoryId() {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/config/AbstractWanPublisherConfig.java]: shouldn't this class force implementers to implement also `getFactoryId` besides `getClassId`?

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/config/AbstractWanPublisherConfig.java]: Good point.

**Comment by mmedenjak** [comment]
@ahmetmircik @blazember changed config from attribute-based to child nodes

File: hazelcast/src/main/java/com/hazelcast/config/AbstractWanPublisherConfig.java (hunk 1 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/config/AbstractWanPublisherConfig.java]: Addressed.

File: hazelcast-spring/src/main/resources/hazelcast-spring-4.0.xsd [hunk 2]
```diff
@@ -2581,127 +2581,34 @@
...
-                        </xs:simpleType>
-                    </xs:attribute>
-                </xs:complexType>
+            <xs:element name="batch-publisher"
+                        type="wan-batch-publisher"
+                        minOccurs="0"
+                        maxOccurs="unbounded">
+                <xs:annotation>
+                    <xs:documentation>
+                        Configuration object for a WAN publisher. A single publisher defines how
```
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast-spring/src/main/resources/hazelcast-spring-4.0.xsd]: Nice catch, thanks!

File: hazelcast/src/main/java/com/hazelcast/internal/management/dto/WanPublisherConfigDTO.java [hunk 3]
```diff
@@ -16,36 +16,28 @@
...
+
 import static com.hazelcast.util.JsonUtil.fromJsonObject;
 import static com.hazelcast.util.JsonUtil.toJsonObject;
 import static com.hazelcast.util.MapUtil.isNullOrEmpty;
 
 /**
- * A JSON representation of {@link WanPublisherConfig}.
+ * A JSON representation of {@link CustomWanPublisherConfig}.
  */
 public class WanPublisherConfigDTO implements JsonSerializable {
```
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/internal/management/dto/WanPublisherConfigDTO.java]: Renamed

File: hazelcast/src/main/resources/hazelcast-full-example.xml [hunk 4]
```diff
@@ -147,12 +147,187 @@
...
-        * <wan-publisher>:
-            Fully qualified class name of WAN Replication implementation implementing WanReplicationEndpoint.
-            Hazelcast Enterprise comes with com.hazelcast.enterprise.wan.impl.replication.WanBatchReplication:
+        * <batch-publisher>:
+            WAN publisher using the built-in com.hazelcast.enterprise.wan.impl.replication.WanBatchReplication.
             Waits until a batch size is reached or a delay time is passed.
-            Please see the `batch.size and batch.max.delay.millis configuration descriptions below.
-        * <wan-consumer>:
+            Please see the batch-size and batch-max-delay-millis configuration descriptions below.
+            * "group-name":
```
[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/resources/hazelcast-full-example.xml]: Ah, ok, leftover from when the config was in attributes, thanks!

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by mmedenjak** [comment]
Thank you a million for the reviews @blazember @ahmetmircik ! I know it was a huge PR and I've already created many cleanup PRs :)

**Pull request merged**